### ISSUE TITLE
[order] Maintain relative imports ordered by path depth when using alphabetize

### ISSUE THREAD

**Issue by fgallardograzio** [benmosher/eslint-plugin-import#1881]
I'd like to suggest that the `order` rule with `alphabetize` enabled should provide a way to keep relative imports ordered by path depth.
This could be configured to be either ascending or descending, to accommodate users preferences.

Currently, using an ascending order, when importing from a folder's index implicitly, `..`, `../..`, etc. get placed before all the non-index imports, ordered in the opposite way, since strictly alphabetically `..` would go before `../..`, but `../x`would go after `../../x`.

IMO implementing this suggested enhancement would lead to a more 'logical', intuitive and user-friendly ordering, where a user would have a better sense of where to find a given import inside a big list of imports, given where it was imported from (relative path depth)  and not based on needing to know whether it was imported from an index file or from a specific file in a given folder.

For example:
IMO this:
```js
import a from '../../a';
import x from '../..';
import b from '../b';
import y from '..';
import c from './c';
import z from '.';
```
or some other variation following the same idea, would make more sense than this (which is the current behavior, after running the autofix):
```js
import z from '.';
import y from '..';
import x from '../..';
import a from '../../a';
import b from '../b';
import c from './c';
```

<details>
<summary>.eslintrc.js</summary>

```js
'import/order': [
    'warn',
    {
        groups: [['parent', 'sibling', 'index']],
        alphabetize: {
            order: 'asc',
        },
    },
],
```

I'm keeping parent, sibling and index in the same group here just as an example, but this suggestion applies equally if they're in separate groups.
</details>

Thanks.

**Comment by ljharb** [comment]
Hmm - this option would only apply to relative imports - but due to custom groups, this seems like it might be tricky to specify.

**Comment by fgallardograzio** [comment]
Imports are already being alphabetized only within their groups, custom or otherwise. Right?

And the code to determine whether a given import is a relative one already exists: `isRelativeToParent`, `isRelativeToSibling`, `isIndex` (which actually just check for path equal to `.` or `..` or path starting with `./`, `.\`, `../` or `..\`).

So we're already restricted to the correct context in which the suggested ordering should be applied, and we have the means to tell if it _should_ be applied, by checking if we're dealing with relative imports or not.

So this would just be a matter of introducing the logic for checking the path depth to the `importsSorter` function, or am i missing something?

And if we wanted to have a setting for this new behavior, I'd say a property in the `alphabetize` object would be ok. Next to `order` and `caseInsensitive`. So it'd be a 'global' setting. I don't think it'd be a good idea to do this on a per-group basis or something like that.

What do you think?
Thanks.

**Comment by ljharb** [comment]
Hmm, that might work.

I'd be happy to review a PR that allowed a per-group setting (not under alphabetize, but a sibling to it) that supported either "ignore", "asc", or "desc" for ordering by relative imports. The docs would have to make it clear that this always took precedence over "alphabetize" when ordering.

(after typing that, I noticed you said you didn't think a per-group basis would be a good idea - can you elaborate on why not? i'm happy with it being *also* global, but any global setting should also be overrideable per group)

**Comment by fgallardograzio** [comment]
Yeah, I mean, I'm not sure I see the value of having different kinds of ordering for different groups. I think it's unlikely someone would want to have, e.g. the `['parent', 'sibling']` group ordered by relative path depth `ignore` and then e.g. have a custom group for, say, graphql files ordered by relative path depth `desc` and another one for assets set to `asc`.

Is that what you meant?

You'd end up with something like this:

<details>

<summary>Code snippet</summary>

```js
import y from '..';
import x from '../..';
import a from '../../a';
import b from './b';

import c from '../../c.graphql';
import d from '../d.graphql';
import e from './e.graphql';

import f from './f.png';
import g from '../g.png';
import h from '../../h.svg';
```

</details>

IMO users would either care about ordering relative imports by path depth in a certain way, or they won't. Just like alphabetization is implemented currently, with the assumption that users want every group ordered either asc or desc, or ignore.

OK, back to the idea of having this be per-group. Then there's the question of how to reference those groups, if we want to have the setting as a sibling to `alphabetize`.
E.g. Using the group string or an array of group strings for the native ones (matching an item in the rule's `groups` array), and for custom groups there's not much of a choice, maybe _string_ equality of the group pattern?

<details>

<summary>Example config</summary>

```js
'import/order': [
    'warn',
    {
        groups: [['parent', 'sibling'], 'object'],
        pathGroups: [
            {
                pattern: '*.graphql',
                patternOptions: { matchBase: true, nocase: true },
                group: 'object',
                position: 'after',
            },
            {
                pattern: '*.{png,svg}',
                patternOptions: { matchBase: true, nocase: true },
                group: 'object',
                position: 'after',
            },
        ],
        alphabetize: {
            order: 'asc',
            caseInsensitive: true,
        },
        orderRelativeImports: [
            {
                group: ['parent', 'sibling'],
                order: 'ignore',
            },
            {
                group: '*.graphql',
                order: 'desc',
            },
            {
                group: '*.{png,svg}',
                order: 'asc',
            },
        ],
    },
],
```

</details>

IMO a config such as that one for `orderRelativeImports` would be very error prone.
What if the user changes `groups`, let's say adding 'index' to the first subgroup: `['parent', 'sibling', 'index']`. Would the first item in the `orderRelativeImports` array simply not match anymore? Or would it do something weird like sort the relative imports as expected until it finds a `'.'` or `'./index'`, and then leave that one unordered?
And what if the user changes a `pathGroup`'s pattern? What if it's an equivalent pattern, like `*.{png,svg}` vs `*.{svg,png}`. It'd be the same but it won't match anymore.

So, all of the above, both the technical aspects as well as the UX, is what led me to think it'd be more reasonable to simply have something like:

```js
'import/order': [
    'warn',
    {
        // ...
        alphabetize: {
            order: 'asc',
            caseInsensitive: true,
        },
        orderRelativeImports: 'asc', // 'asc'|'desc'|'ignore'
    },
],
```

Thoughts?

----

One more note, when I initially thought of this proposal I thought of it as a particular case for the already implemented alphabetization, like the same way you could say 'alphabetize imports in ascending order, _but order numbers by their numeric value, not their character code_', I thought '..., _but keep relative paths ordered by depth_'.
That's why I said we could have the option as a property inside the `alphabetize` object. And since alphabetization is 'global', not per-group, this would've been the same.

But having it as a standalone option, as a sibling to alphabetize, and working independently, sounds very good as well.

**Comment by namnm** [comment]
How's about this one: https://github.com/lydell/eslint-plugin-simple-import-sort
